Documentation for roles v2. (#868)
osohq/oso: docs/content/python/new-roles/examples.md

---
title: Role modeling by example
weight: 3
description: >
    Let's see how to build common role models with SQLAlchemy roles.
---

# Role modeling by example

In this guide, we'll cover a few different role models in the context of
the [GitClub example application](https://github.com/osohq/gitclub-sqlalchemy-flask-react).

{{% callout "Read first" "blue" %}}
 - [Getting started](getting-started)
{{% /callout %}}

## Multiple role types: Repository roles

> In the [Getting Started](getting-started) guide, we only discussed a role for a single
> resource type. However, many apps will have multiple levels of roles. It

Reviewer: ```suggestion
In the [Getting Started](getting-started) guide, we only discussed a role for a single
resource type (`Org`). However, many apps will have multiple levels of roles. It
```

is common to have roles associated with an organization object, like

> owner and member, and more granular roles associated with sub-roles. In

Reviewer: ```suggestion
"owner" and "member", and more granular roles associated with resources within an organization/tenant (we'll refer to these as "child resources"). In
```

> the GitClub example application, the Organization model has a member and
> owner role. Repositories have reader and writer roles.

Reviewer: ```suggestion
the GitClub example application, the Organization model has "member" and
"owner" roles. The Repository model has "reader" and "writer" roles.
```

To define a role for another resource type, we add another `resource`
rule to our policy.

{{< literalInclude
    path="examples/gitclub-sqlalchemy-flask-react/backend/app/authorization.polar"
    from="docs: begin-repo-resource"
    to="docs: end-repo-resource"
    gitHub="https://github.com/osohq/gitclub-sqlalchemy-flask-react"
    linenos=true
>}}

This defines two roles on repository, `"repo_read"` and `"repo_write"`.
This allows us to assign users directly to repositories without
giving them a role in the entire organization.

Now, we can use `OsoRoles.assign_role` to assign role on
repositories in addition to organizations.

> ## Grant access to sub-resources with implied roles

Reviewer: ```suggestion
## Grant access to child resources with implied roles
```
To keep terminology consistent

Often a role on a parent resource will grant access to all child
resources. In GitClub, a member of an organization is granted read access to
all repositories within that organization.

> We can model this access control model with **cross resource implied
> roles**.

Reviewer: ```suggestion
We can model this access control model with **cross-resource implied
roles**.
```

repository's `org` field.

When Oso evalutes the policy, it uses the `org` relationship defined
on our model:

{{< literalInclude
    path="examples/gitclub-sqlalchemy-flask-react/backend/app/models.py"
    from="docs: begin-repo-model"
    to="docs: end-repo-model"
    gitHub="https://github.com/osohq/gitclub-sqlalchemy-flask-react"
    linenos=true
    hlOpts="hl_lines=9"
>}}

Then, we **imply** a role on the child resource from our parent resource
definition.

{{< literalInclude
    path="examples/gitclub-sqlalchemy-flask-react/backend/app/authorization.polar"
    from="docs: begin-org-resource"
    to="docs: end-org-resource"
    gitHub="https://github.com/osohq/gitclub-sqlalchemy-flask-react"
    linenos=true
    hlOpts="hl_lines=6"
    >}}

The `"repo_read"` entry in `implies: ` for `org_member` gives organization
members the `repo_read` role on all repositories within the
organization.

{{% callout "Parent limitations" "blue" %}}

Currently a `parent` rule body must be of the form:

```polar
parent(child: ChildType, parent: ParentType) if
    child.attribute = parent;
```

This is sufficient to model resources with single parents.

{{% /callout %}}

> ## Cross resource permissions

Reviewer: ```suggestion
## Cross-resource permissions
```

Sometimes, we may want to grant permissions on child resources but do
not need that resource to have its own roles.

An example in GitClub is issues. Repository writers can read issues, but
repository readers cannot.

First, we define the issue resource in our policy. Even though an issue

> doesn't have roles, we still define it to create the actions.

Reviewer: ```suggestion
doesn't have roles, we still define an issue resource to declare the actions users can take on issues.
```

{{< literalInclude
    path="examples/gitclub-sqlalchemy-flask-react/backend/app/authorization.polar"
    from="docs: begin-issue-resource"
    to="docs: end-issue-resource"
    gitHub="https://github.com/osohq/gitclub-sqlalchemy-flask-react"
    linenos=true
    >}}

Notice the last argument to `resource` is `_` instead of `roles`. This
indicates it is unused.

> Then, we just assign the issue action as a permission in a role on the
> parent. Notice the `"issue:read"` permission for `repo_write` below:

Reviewer: ```suggestion
Now, we can assign the issue action to a repository role. Notice the `"issue:read"` permission for `repo_write` below:
```
